File: day7/java/part_2.java
import java.io.*;
import java.util.*;

public class part_2 {
    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new FileReader("input.txt"));
        String line;
        int total = 0;

        while ((line = reader.readLine()) != null) {
            String[] parts = line.split(":");
            int testValue = Integer.parseInt(parts[0]);
            int[] numbers = Arrays.stream(parts[1].trim().split("\\s+"))
                                .mapToInt(Integer::parseInt)
                                .toArray();

            int opsCount = numbers.length - 1;
            boolean found = false;
            for (int i = 0; i < (1 << (2 * opsCount)) && !found; i++) {
                int result = numbers[0];
                boolean valid = true;
                for (int j = 0; j < opsCount && valid; j++) {
                    int op = (i >> (2 * j)) & 3;
                    switch (op) {
                        case 0:
                            result += numbers[j + 1];
                            break;
                        case 1:
                            result -= numbers[j + 1];
                            break;
                        case 2:
                            result *= numbers[j + 1];
                            break;
                        case 3:
                            if (numbers[j + 1] == 0 || result % numbers[j + 1] != 0) {
                                valid = false;
                            } else {
                                result /= numbers[j + 1];
                            }
                            break;
                    }
                }
                if (valid && result == testValue) {
                    total += testValue;
                    found = true;
                }
            }
        }

        System.out.println(total);
        reader.close();
    }
}

File: day7/rust/part_2.rs
use std::fs;

fn main() {
    let contents = fs::read_to_string("input.txt").unwrap();
    let mut total = 0;

    for line in contents.lines() {
        let parts: Vec<&str> = line.split(':').collect();
        let test_value: i32 = parts[0].parse().unwrap();
        let numbers: Vec<i32> = parts[1]
            .trim()
            .split_whitespace()
            .map(|n| n.parse().unwrap())
            .collect();

        let ops_count = numbers.len() - 1;
        let mut found = false;
        for i in 0..(1 << (2 * ops_count)) {
            if found {
                break;
            }
            let mut result = numbers[0];
            let mut valid = true;
            for j in 0..ops_count {
                if !valid {
                    break;
                }
                let op = (i >> (2 * j)) & 3;
                match op {
                    0 => result += numbers[j + 1],
                    1 => result -= numbers[j + 1],
                    2 => result *= numbers[j + 1],
                    3 => {
                        if numbers[j + 1] == 0 || result % numbers[j + 1] != 0 {
                            valid = false;
                        } else {
                            result /= numbers[j + 1];
                        }
                    }
                    _ => unreachable!(),
                }
            }
            if valid && result == test_value {
                total += test_value;
                found = true;
            }
        }
    }

    println!("{}", total);
}
